opencog/distvalue/CoverTree.cc

```diff
@@ -0,0 +1,576 @@
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <sstream>
+#include <cmath>
+
+#include <opencog/util/exceptions.h>
+#include <opencog/util/numeric.h>
+
+#include <opencog/distvalue/CoverTree.h>
+
+using namespace std;
+using namespace opencog;
+
+//CoverTreeNode
+
+template <typename val_t>
+bool CoverTreeNode<val_t>::operator!=(const CoverTreeNode<val_t>& other) const
+{
+	return !(*this == other);
+}
+
+template <typename val_t>
+bool CoverTreeNode<val_t>::operator==(const CoverTreeNode<val_t>& other) const
+{
+	if (pos != other.pos)
+		return false;
+
+	if (value != other.value)
+		return false;
+
+	for (auto elem : children)
+	{
+		auto res = std::find(other.children.begin(),other.children.end(), elem);
+
+		if (res == other.children.end())
+			return false;
+
+		if (elem != *res)
+			return false;
+	}
+	return true;
+}
+
+//CoverTree
+
+template <typename val_t>
+CoverTree<val_t>::CoverTree()
+	: _root_idx(-1) , _root_level(0) , _total_count(0) , _dims(1) {}
+
+template <typename val_t>
+CoverTree<val_t>::CoverTree(int dims)
+	:  _root_idx(-1) , _root_level(0) , _total_count(0) , _dims(dims) {}
+
+template <typename val_t>
+CoverTree<val_t>::CoverTree(CoverTreeNode<val_t> n,int dims)
+	:  _root_idx(-1) , _root_level(0) , _dims(dims) {
+	_nodes.push_back(n);
+	_root_idx = _nodes.size() - 1;
+    _total_count = get_count(n.value);
+}
+
+template <typename val_t>
+double CoverTree<val_t>::dist(const CoverTreeNode<val_t> & n1,
+                              const CoverTreeNode<val_t> & n2)
+{
+	return ::dist(n1.pos,n2.pos);
+}
+
+template <typename val_t>
+double CoverTree<val_t>::covdist(int level)
+{
+	return pow(2,level);
+}
+
+//TODO: Cache This?
+template <typename val_t>
+double CoverTree<val_t>::maxdist(const CoverTreeNode<val_t> & n) const
+{
+	double max = 0;
+	for (int c : n.children)
+	{
+		double tmp = dist(n,_nodes[c]);
+		if (tmp > max)
+			max = tmp;
+	}
+	return max;
+}
+
+template <typename val_t>
+int CoverTree<val_t>::popLeaf(CoverTreeNode<val_t> & n)
+{
+	int c_id = n.children.back();
+	CoverTreeNode<val_t> & c = _nodes[c_id];
+	if (c.children.size() == 0)
+	{
+		n.children.pop_back();
+		return c_id;
+	}
+	return popLeaf(c);
+}
+
+template <typename val_t>
+const CoverTreeNode<val_t> *
+CoverTree<val_t>::findNearestNeighbor(const DVec & pos) const
+{
+	const CoverTreeNode<val_t> & x = CoverTreeNode<val_t>(pos,val_t());
```

Comment:
Are you sure you can do that without risking that `x` get deallocated before used below?

I know you can pass the constructor directly to `findNearestNeighbor_`, so by that logic it should be OK, but I'm not 100%, C++ is a weird beast sometimes.

Reply:
The call to findNearestNeighbor_ is inside the scope of findNearestNeighbor were x has been created so it should work fine. And i have not noticed any issues in this regard.